examples/Tutorial 4 - 3SAT Solver.ipynb

```diff
@@ -0,0 +1,268 @@
+    "        self.terms = terms\n",
+    "\n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this clause with the value of the variables as specified in `values`.\n",
+    "        Returns `True` if and only for any `term`, `term.eval(values)` returns `True`.\n",
+    "        \"\"\"\n",
+    "        if len(self.terms) != 3:\n",
+    "            raise ValueError(\"A 3SAT clause must have exactly three terms.\")\n",
+    "        for term in self.terms:\n",
+    "            if term.eval(values):\n",
+    "                return True\n",
+    "        return False\n",
+    "        pass\n",
+    "    \n",
+    "    \n",
+    "\n",
+    "\n",
+    "class ThreeSATInstance:\n",
+    "    \"\"\"\n",
+    "    A 3SAT instance, i.e. C1 AND C2 AND C3  AND Cn where each Ci is a 3SAT Clause.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    clauses: list[ThreeSATClause]\n",
+    "    _variables: list[str]\n",
+    "    \n",
+    "    def __init__(self, clauses):\n",
+    "        self.clauses = clauses\n",
+    "        self._variables = []\n",
+    "        \n",
+    "        for clause in self.clauses:\n",
+    "            for term in clause.terms:\n",
+    "                if term.variable not in self._variables:\n",
+    "                    self._variables.append(term.variable)\n",
+    "\n",
+    "\n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this instance with the value of the variables as specified in `values`.\n",
+    "\n",
+    "        Returns `True` if and only for each `clause`, `clause.eval(values)` returns `True`.\n",
+    "        \"\"\"\n",
+    "        for clause in self.clauses:\n",
+    "            if not clause.eval(values):\n",
+    "                return False\n",
+    "        return True\n",
+    "        \n",
+    "    \n",
+    "\n",
+    "    def __len__(self) -> int:\n",
+    "        \"\"\"\n",
+    "        The number of clauses in this instance.\n",
+    "        \"\"\"\n",
+    "        return len(self.clauses)\n",
+    "    \n",
+    "\n",
+    "    def compile_to_mis(self) -> MISInstance:\n",
+    "        \"\"\"\n",
+    "            Compile this instance of 3SAT into an instance of MIS.\n",
+    "        \"\"\"\n",
+    "        \n",
+    "        # keep track of the nodes of each variable (and complementary)\n",
+    "        nodes = defaultdict(lambda : []) # set the default value of a key to an empty list\n",
+    "        \n",
+    "        graph = nx.Graph()\n",
+    "        for i in range(len(self.clauses)):\n",
+    "            graph.add_edges_from([(3*i,3*i+1), (3*i+1, 3*i+2), (3*i, 3*i+2)])\n",
+    "            # add edges between each node and its complementaries\n",
+    "            for j in range(3):\n",
+    "                for node in nodes[self.clauses[i].terms[j].variable + (\"F\" if self.clauses[i].terms[j].positive else \"T\")]:\n",
```

Comment:
Not a big fan of this large expression copied and copied. Perhaps make it a lambda or a private method?